pkg/autoscaler/autoscaler.go

```diff
@@ -254,295 +257,409 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
-		// this is needed due to the autoscaler tests requiring singleton
-		// deployments of the ClusterAutoscaler.
-		caName := clusterAutoscaler.GetName()
-		Expect(deleteObject(caName, cleanupObjects[caName])).Should(Succeed())
-		delete(cleanupObjects, caName)
-		Eventually(func() bool {
-			By(fmt.Sprintf("Waiting for ClusterAutoscaler to delete."))
-			if ca, err := framework.GetClusterAutoscaler(client, caName); ca == nil && err != nil {
-				return true
+			for _, machineSet := range transientMachineSets {
+				By(fmt.Sprintf("Waiting for MachineSet %s replicas to scale out", machineSet.GetName()))
 			}
-			return false
-		}, framework.WaitLong, pollingInterval).Should(BeTrue())
+			Eventually(func() (bool, error) {
+				for _, machineSet := range transientMachineSets {
+					current, err := framework.GetMachineSet(client, machineSet.GetName())
+					if err != nil {
+						return false, err
+					}
+					if *current.Spec.Replicas != expectedReplicas {
+						return false, nil
+					}
+				}
+				return true, nil
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+
+			By("Waiting for all Machines in MachineSets to enter Running phase")
+			framework.WaitForMachineSet(client, transientMachineSets[0].GetName())
+			framework.WaitForMachineSet(client, transientMachineSets[1].GetName())
+
+			By("Deleting the workload")
+			Expect(deleteObject(workload.Name, cleanupObjects[workload.Name])).Should(Succeed())
+			delete(cleanupObjects, workload.Name)
+
+			for _, machineSet := range transientMachineSets {
+				By(fmt.Sprintf("Waiting for MachineSet %s replicas to scale in", machineSet.GetName()))
+			}
+			expectedReplicas = 1
+			Eventually(func() (bool, error) {
+				for _, machineSet := range transientMachineSets {
+					current, err := framework.GetMachineSet(client, machineSet.GetName())
+					if err != nil {
+						return false, err
+					}
+					if current.Status.Replicas != expectedReplicas {
+						return false, nil
+					}
+				}
+				return true, nil
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+
+			for _, machineSet := range transientMachineSets {
+				machines, err := framework.GetMachinesFromMachineSet(client, machineSet)
+				Expect(err).NotTo(HaveOccurred())
+				for _, machine := range machines {
+					By(fmt.Sprintf("Checking Machine %s for %s annotation", machine.Name, machineDeleteAnnotationKey))
+				}
+			}
+			Eventually(func() (bool, error) {
+				for _, machineSet := range transientMachineSets {
+					machines, err := framework.GetMachinesFromMachineSet(client, machineSet)
+					if err != nil {
+						return false, err
+					}
+					for _, machine := range machines {
+						if machine.ObjectMeta.Annotations == nil {
+							continue
+						}
+						if _, exists := machine.ObjectMeta.Annotations[machineDeleteAnnotationKey]; exists {
+							return false, nil
+						}
+					}
+				}
+				return true, nil
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+		})
```

Comment:
This seems a bit odd to me, we iterate through the list once to print that we are checking for the annotation, then iterate throught the list again for actually checking, why not have the eventually immediately after the by and just have it check for the one machine having the annotation?

Reply:
it does look a little odd, i agree. my reasoning here was to print the names of the machines we are going to be watching for removal of annotations. but, i didn't want to print continuously through the `Eventually` clause (since it can take a few loops to finish).

i can either refactor this or add some comments to help explain, wdyt?

Reply:
Does this sort of thing not work?
```suggestion
			for _, machineSet := range transientMachineSets {
				machines, err := framework.GetMachinesFromMachineSet(client, machineSet)
				Expect(err).NotTo(HaveOccurred())
				for _, machine := range machines {
					By(fmt.Sprintf("Checking Machine %s for %s annotation", machine.Name, machineDeleteAnnotationKey))
					Eventually(func() (bool, error) {
					    m := &Machine{}
					    err := client.Get(ctx, client.ObjectKey{Namespace: machine.Namespace, Name: machine.Name}, m)
					    if err != nil {
					      return false, err
					    }
					    
					    if m.ObjectMeta.Annotations == nil {
					      return true, nil
					    }
					    if _, exists := machine.ObjectMeta.Annotations[machineDeleteAnnotationKey]; exists {
				              return false, nil
					    }
					    return true, nil
					})
				}
			}
```

Reply:
this makes sense to me, let me re-work this bit a little